savings-account/src/lib.rs

```diff
@@ -2,8 +2,390 @@
+
+        self.burn_tokens(&borrow_token_id, borrow_token_nonce, borrow_token_amount)?;
+
+        let caller = self.blockchain().get_caller();
+        let extra_stablecoins_paid = stablecoin_amount - &total_stablecoins_needed;
+        if extra_stablecoins_paid > 0 {
+            self.send().direct(
+                &caller,
+                &stablecoin_token_id,
+                0,
+                &extra_stablecoins_paid,
+                &[],
+            );
+        }
+
+        // TODO: Maybe automatically claim staking rewards here for one last time?
+
+        let liquid_staking_token_id = self.liquid_staking_token_id().get();
+        self.send().direct(
+            &caller,
+            &liquid_staking_token_id,
+            borrow_metadata.liquid_staking_token_nonce,
+            borrow_token_amount,
+            &[],
+        );
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn withdraw(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_nonce] payment_nonce: u64,
+        #[payment_amount] payment_amount: Self::BigUint,
+    ) -> SCResult<()> {
+        self.require_lend_token_issued()?;
+
+        let stablecoin_token_id = self.stablecoin_token_id().get();
+        let lend_token_id = self.lend_token_id().get();
+        require!(
+            payment_token == lend_token_id,
+            "May only pay with lend tokens"
+        );
+
+        let mut lend_metadata = self.lend_metadata(payment_nonce).get();
+        lend_metadata.amount_in_circulation -= &payment_amount;
+
+        if lend_metadata.amount_in_circulation == 0 {
+            self.lend_metadata(payment_nonce).clear();
+        } else {
+            self.lend_metadata(payment_nonce).set(&lend_metadata);
+        }
+
+        let borrowed_amount = self.borrowed_amount().get();
+        let stablecoin_reserves = self.get_reserves(&stablecoin_token_id);
+        let current_utilisation =
+            self.compute_capital_utilisation(&borrowed_amount, &stablecoin_reserves);
+
+        let pool_params = self.pool_params().get();
+        let borrow_rate = self.compute_borrow_rate(
+            &pool_params.base_borrow_rate,
+            &pool_params.borrow_rate_under_opt_factor,
+            &pool_params.borrow_rate_over_opt_factor,
+            &pool_params.optimal_utilisation,
+            &current_utilisation,
+        );
+        let deposit_rate = self.compute_deposit_rate(
+            &current_utilisation,
+            &borrow_rate,
+            &pool_params.reserve_factor,
+        );
+
+        let withdraw_amount =
+            self.compute_withdrawal_amount(&payment_amount, lend_metadata.timestamp, &deposit_rate);
+        let sc_stablecoin_balance = self.blockchain().get_sc_balance(&stablecoin_token_id, 0);
+
+        if withdraw_amount > sc_stablecoin_balance {
+            /* TODO:
+                Try convert EGLD to X worth of stablecoins (where X =  withdraw_amount - sc_balance)
```

Comment:
No. When the claimRewards happen - than the trade will happen - of changing EGLD to StableCoin on the open market. Thus selling eGLD and buying stable coin. There you compute if the earned stablecoins are enough to pay out the lenders - based of the defined constant percentage of winning. What is on top is put into a reserve pool